Cargo.lock

```diff
@@ -207,6 +209,17 @@ name = "fixedbitset"
 version = "0.1.8"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 
+[[package]]
+name = "flate2"
+version = "1.0.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "futures 0.1.18 (registry+https://github.com/rust-lang/crates.io-index)",
+ "libc 0.2.36 (registry+https://github.com/rust-lang/crates.io-index)",
+ "miniz-sys 0.1.10 (registry+https://github.com/rust-lang/crates.io-index)",
```

Comment:
I'm not convinced we should depend on miniz.

Why not use deflate-rs instead of flate2? We only compress (deflate), never decompress (inflate), and only for metrics, so we don't need maximum compression or maximum performance. I'd rather use the pure Rust implementation (first preference) or zlib, unless there's a compelling reason to use the miniz implementation.

Reply:
I chose `flate2` initially as it had `tokio-io` support, so that the compression task could be managed by the reactor rather than spawning a thread. If you don't think that's important, I could rewrite with `deflate-rs`. Alternatively, `flate2` also has a feature flag to use the pure-Rust `miniz-oxide` rather than `miniz-sys`, but I'm not sure how stable that is currently.

Reply:
Currently, the whole thing is buffered and compressed in 1 pass, so non-blocking isn't really being used, right?

Reply:
> Alternatively, flate2 also has a feature flag to use the pure-Rust miniz-oxide rather than miniz-sys, but I'm not sure how stable that is currently.

My understanding is that miniz-oxide is a Rust implementation of a C API. Would flate2 then be doing safe -> unsafe -> safe calls using miniz-oxide?

It seems to me that deflate-rs is the best choice since it's pure Rust and its developer seems interested in cooperating with others to improve it. It's not used directly by a lot of crates, but it is used by the `png` crate that servo uses.

Reply:
@briansmith et al., I rewrote this branch to depend on `deflate-rs` in50ebaac3aff976857cc3bba5a92698b345caf61f. 

`flate2` is still used in tests, as `deflate` only implements compression, not decompression, and we need to be able to decompress gzipped metrics responses to assert that they're correct. However, it's only listed as a `dev-dependency` in Cargo.toml. I'd like to remove the dependency on `flate2` entirely, but that would require `deflate-rs` adding support for decompression.

Reply:
> late2 is still used in tests, as deflate only implements compression, not decompression, and we need to be able to decompress gzipped metrics responses to assert that they're correct. However, it's only listed as a dev-dependency in Cargo.toml. I'd like to remove the dependency on flate2 entirely, but that would require deflate-rs adding support for decompression.

There is [inflate](https://crates.io/crates/inflate) that seems much more lightweight in terms of dependencies and which doesn't depend on any native code.

Since flate2 is only a dev-dependency, the flate2 dependency isn't a blocker. But IMO we should switch to the inflate crate or another pure-rust crate either in this PR or in a follow-up soon after.

Reply:
> There is inflate that seems much more lightweight in terms of dependencies and which doesn't depend on any native code.

Yeah, I was mainly thinking that it would be nice to not add a dev dependency for decompression at all, but it should't be much work to rewrite the tests to use `inflate` rather than `flate2`.

Reply:
@briansmith, I took a look at switching the dev dependency to `inflate` rather than `flate2`, but it appears that `inflate` only supports decompressing DEFLATE and ZLIB compression, not GZIP. Prometheus specifically requires GZIP compression, so we need to be able to test that encoding.

Reply:
:( I suspected that might be the case since its only dependency was adler32. Ok, flate2 then.